drv/sprot-api/src/lib.rs

```diff
@@ -333,232 +422,238 @@ impl From<u8> for MsgType {
     }
 }
 
-#[derive(Copy, Clone, Serialize, Deserialize, SerializedSize)]
-#[repr(C, packed)]
-struct MsgHeader {
-    protocol: Protocol,
-    msgtype: MsgType,
-    len: u16,
+/// A builder/serializer for messages that wraps the transmit buffer
+///
+/// Each public method returns the serialized buffer that can be sent on the
+/// wire.
+pub struct TxMsg {
+    buf: [u8; BUF_SIZE],
 }
 
-#[derive(
-    Copy, Clone, Eq, PartialEq, Serialize, Deserialize, SerializedSize,
-)]
-#[repr(C)]
-pub enum UpdateRspKind {
-    Ok = 1,
-    Value = 2,
-    Error = 3,
-    Unknown = 0xff,
-}
+impl TxMsg {
+    pub fn new() -> TxMsg {
+        TxMsg { buf: [0; BUF_SIZE] }
+    }
 
-impl From<u8> for UpdateRspKind {
-    fn from(kind: u8) -> Self {
-        match kind {
-            kind if kind == (UpdateRspKind::Ok as u8) => UpdateRspKind::Ok,
-            kind if kind == (UpdateRspKind::Value as u8) => {
-                UpdateRspKind::Value
-            }
-            kind if kind == (UpdateRspKind::Error as u8) => {
-                UpdateRspKind::Error
-            }
-            _ => UpdateRspKind::Unknown,
+    pub fn as_slice(&self) -> &[u8] {
+        &self.buf[..]
+    }
+
+    pub fn as_mut(&mut self) -> &mut [u8] {
+        &mut self.buf[..]
+    }
+
+    pub fn payload_mut(&mut self) -> &mut [u8] {
```

Comment:
These three methods feel a little questionable to me, but maybe they're fine. Just thinking out loud - `TxBuf` is sort of a higher-level thing to assemble messages (e.g., `error_rsp()`, `no_payload()`, and `from_lease()` all build a message without the caller having to deal with the details of the buffer), but these methods (and `from_existing()`) require the caller to have relatively intricate knowledge of how the message is assembled to use correctly (either to write to the correct place, or to know where to stop reading).

Reply:
I totally agree @jgallagher. This was sort of a compromise with the code as it already existed. However, I think I can tighten this up some more.